### PR TITLE
Track the missing dependencies on SPDX documents

### PR DESCRIPTION
**Pull request by quaresmajose**
This handler will generate an ERROR when there are 'do_deploy' dependencies without setting the DEPENDS.
One of the side effect of this issues is SPDX documents not including the correct dependencies.
```
/
| ERROR: lmp-boot-firmware-0-r0 do_install: Task 'do_install' depends on 'virtual/bootloader:do_deploy' but 'virtual/bootloader' is not in DEPENDS
| ERROR: lmp-boot-firmware-0-r0 do_install: Task 'do_install' depends on 'virtual/trusted-firmware-a:do_deploy' but 'virtual/trusted-firmware-a' is not in DEPENDS
\
```

Explicit adds the dependencies of the 'do_deploy' in the task[depends].
With this patch the missed dependencies are now in the SPDX documents of the package as well as in the final SPDX of the combined image.

The following example is for the lmp-base-console-image on stm32mp15-disco machine:
```
/
| grep -E "u-boot|tf-a" deploy/images/stm32mp15-disco/lmp-base-console-image-stm32mp15-disco.spdx.index.json
| deploy/images/stm32mp15-disco/lmp-base-console-image-stm32mp15-disco.spdx.index.json:      "documentNamespace": "http://spdx.org/spdxdoc/recipe-tf-a-fio-546ca65d-3ceb-5456-93cd-e8acffa32a17",
| deploy/images/stm32mp15-disco/lmp-base-console-image-stm32mp15-disco.spdx.index.json:      "filename": "recipe-tf-a-fio.spdx.json",
| deploy/images/stm32mp15-disco/lmp-base-console-image-stm32mp15-disco.spdx.index.json:      "documentNamespace": "http://spdx.org/spdxdoc/recipe-tf-a-tools-native-52687ea6-9f82-5a2d-a8ab-f4e90f31e6fd",
| deploy/images/stm32mp15-disco/lmp-base-console-image-stm32mp15-disco.spdx.index.json:      "filename": "recipe-tf-a-tools-native.spdx.json",
| deploy/images/stm32mp15-disco/lmp-base-console-image-stm32mp15-disco.spdx.index.json:      "documentNamespace": "http://spdx.org/spdxdoc/recipe-u-boot-fio-cfb5b3c2-fc87-5d33-a223-5b46ae9bb0f5",
| deploy/images/stm32mp15-disco/lmp-base-console-image-stm32mp15-disco.spdx.index.json:      "filename": "recipe-u-boot-fio.spdx.json",
\
```

### CODE REVIEW

**Comment by ricardosalveti** [comment]
Please extend the commit messages to avoid patches with just the subject line (add more details as part of the commit body message).

File: meta-lmp-base/recipes-bsp/lmp-boot-firmware/lmp-boot-firmware.bb [hunk 1]
```diff
@@ -54,16 +54,15 @@ do_install() {
             for file in `ls ${D}${nonarch_base_libdir}/${FIRMWARE_DEPLOY_DIR}/`; do
                 version="${version}-`md5sum ${D}${nonarch_base_libdir}/${FIRMWARE_DEPLOY_DIR}/${file} | cut -d' ' -f1`"
             done
-
-	    # limit the length of version
-	    version="`echo ${version} | md5sum | cut -d' ' -f1`"
+            # limit the length of version
+            version="`echo ${version} | md5sum | cut -d' ' -f1`"
```
[the remark this answers is not in the document]
Comment (quaresmajose) [reply meta-lmp-base/recipes-bsp/lmp-boot-firmware/lmp-boot-firmware.bb]: yes

**Comment by MrCry0** [comment]
in the commit msg: `s/combinad/combined/` ?

**Comment by quaresmajose** [comment]
> in the commit msg: `s/combinad/combined/` ?

fixed! thanks

**Review by MrCry0 — APPROVED** [review]
LGTM

File: meta-lmp-base/recipes-bsp/lmp-boot-firmware/lmp-boot-firmware.bb [hunk 2]
```diff
@@ -86,3 +86,9 @@ FILES:${PN} = "\
 	${nonarch_base_libdir}/firmware \
 	${nonarch_base_libdir}/ostree-boot \
 "
+
+python () {
+    # we need to set the DEPENDS as well to produce valid SPDX documents
+    for tasks in d.getVarFlag('do_install', 'depends').split():
+        d.appendVar('DEPENDS', " %s" % tasks.split(":")[0])
```
Comment (angolini) [inline meta-lmp-base/recipes-bsp/lmp-boot-firmware/lmp-boot-firmware.bb:93]: Setting DEPENDS usually goes on the top of the file. Is there any reason it must be at the bottom in this case?

**Review by angolini — APPROVED** [review]
LGTM

File: meta-lmp-base/recipes-bsp/lmp-boot-firmware/lmp-boot-firmware.bb (hunk 2 above)
Comment (quaresmajose) [reply meta-lmp-base/recipes-bsp/lmp-boot-firmware/lmp-boot-firmware.bb:93]: Nothing special, only because it is using a non comam way anonymous python.

**Comment by quaresmajose** [comment]
validated on https://github.com/foundriesio/lmp-manifest/pull/318

File: meta-lmp-base/classes/lmp-staging.bbclass [hunk 3]
```diff
@@ -38,3 +38,38 @@ def lmp_sstate_checkhashes(sq_data, d, **kwargs):
...
+python check_deployed_depends() {
+    d = e.data
+
+    taskname = d.getVar('BB_RUNTASK')
+    pn = d.getVar('PN')
+    pn = [pn, '%s:%s' % (pn, taskname)]
+    excludes = (d.getVar('LMPSTAGING_DEPLOYED_CHECK_SKIP') or '').split()
+    for s in excludes:
+        if s in pn:
+            #bb.warn("skip '%s' deployed dependencies check" % s)
```
Comment (ricardosalveti) [inline meta-lmp-base/classes/lmp-staging.bbclass:57]: Drop this line.

File: meta-lmp-base/classes/lmp-staging.bbclass [hunk 4]
```diff
@@ -38,3 +38,38 @@ def lmp_sstate_checkhashes(sq_data, d, **kwargs):
...
+            bb.debug(1, "skip '%s' deployed dependencies check" % s)
+            return
+
+    depends = (d.getVarFlag(taskname, 'depends') or '').split()
+    # remove duplicates
+    depends = [*set(depends)]
+    for depend in depends:
+        recipe, task = depend.split(':')
+        if task == 'do_deploy' and recipe not in d.getVar('DEPENDS'):
+            bb.error("Task '%s' depends on '%s' but '%s' is not in DEPENDS" % (taskname, depend, recipe))
```
Comment (ricardosalveti) [inline meta-lmp-base/classes/lmp-staging.bbclass:67]: I think it is better for us to warn by default instead of calling an error here, as this might have side effects in customer related recipes and cause certain builds to fail.

While we do want to have SPDX in a correct form, we shouldn't just yet enforce this on every recipe (we can have as a warn first and on another release move to error, for example).

File: meta-lmp-base/recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_%.bbappend [hunk 5]
```diff
@@ -17,3 +17,10 @@ do_install:append() {
 
 INITRAMFS_RECOVERY_IMAGE ?= ""
 do_install[depends] += "virtual/kernel:do_deploy ${@['${INITRAMFS_RECOVERY_IMAGE}:do_image_complete', ''][d.getVar('INITRAMFS_RECOVERY_IMAGE') == '']}"
+
+INHIBIT_DEFAULT_DEPS = "1"
```
Comment (ricardosalveti) [inline meta-lmp-base/recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_%.bbappend:21]: Split this on in a different commit.

Comment (quaresmajose) [reply meta-lmp-base/recipes-sota/ostree-kernel-initramfs/ostree-kernel-initramfs_%.bbappend]: done

File: meta-lmp-base/classes/lmp-staging.bbclass (hunk 4 above)
Comment (quaresmajose) [reply meta-lmp-base/classes/lmp-staging.bbclass]: I put a new commit [25a255e](https://github.com/foundriesio/meta-lmp/pull/1109/commits/25a255ec5211dff0c185ddcfd8047959b256be23) wih this that is simpler to reference later.

**Review by ricardosalveti — APPROVED** [review]
LGTM
